devito/pointdata.py

```diff
@@ -0,0 +1,197 @@
+
+            # Allocate and copy coordinate data
+            self.coordinates = DenseData(name='%s_coords' % self.name,
+                                         dimensions=[self.indices[1], d],
+                                         shape=(self.npoint, self.ndim))
+            coordinates = kwargs.get('coordinates', None)
+            if coordinates is not None:
+                self.coordinates.data[:] = coordinates[:]
+
+    def __new__(cls, *args, **kwargs):
+        nt = kwargs.get('nt')
+        npoint = kwargs.get('npoint')
+        kwargs['shape'] = (nt, npoint)
+
+        return DenseData.__new__(cls, *args, **kwargs)
+
+    @classmethod
+    def _indices(cls, **kwargs):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        :return: indices used for axis.
+        """
+        dimensions = kwargs.get('dimensions', None)
+        return dimensions or [t, p]
+
+    @property
+    def coefficients(self):
+        """Symbolic expression for the coefficients for sparse point
+        interpolation according to:
+        https://en.wikipedia.org/wiki/Bilinear_interpolation.
+
+        :returns: List of coefficients, eg. [b_11, b_12, b_21, b_22]
+        """
+        # Grid indices corresponding to the corners of the cell
+        x1, y1, z1, x2, y2, z2 = symbols('x1, y1, z1, x2, y2, z2')
+        # Coordinate values of the sparse point
+        px, py, pz = self.point_symbols
+        if self.ndim == 2:
+            A = Matrix([[1, x1, y1, x1*y1],
+                        [1, x1, y2, x1*y2],
+                        [1, x2, y1, x2*y1],
+                        [1, x2, y2, x2*y2]])
+
+            p = Matrix([[1],
+                        [px],
+                        [py],
+                        [px*py]])
+
+        elif self.ndim == 3:
+            A = Matrix([[1, x1, y1, z1, x1*y1, x1*z1, y1*z1, x1*y1*z1],
+                        [1, x1, y2, z1, x1*y2, x1*z1, y2*z1, x1*y2*z1],
+                        [1, x2, y1, z1, x2*y1, x2*z1, y2*z1, x2*y1*z1],
+                        [1, x1, y1, z2, x1*y1, x1*z2, y1*z2, x1*y1*z2],
+                        [1, x2, y2, z1, x2*y2, x2*z1, y2*z1, x2*y2*z1],
+                        [1, x1, y2, z2, x1*y2, x1*z2, y2*z2, x1*y2*z2],
+                        [1, x2, y1, z2, x2*y1, x2*z2, y1*z2, x2*y1*z2],
+                        [1, x2, y2, z2, x2*y2, x2*z2, y2*z2, x2*y2*z2]])
+
+            p = Matrix([[1],
+                        [px],
+                        [py],
+                        [pz],
+                        [px*py],
+                        [px*pz],
+                        [py*pz],
+                        [px*py*pz]])
+        else:
+            error("Interpolation coefficients not implemented for "
+                  "%d dimensions." % self.ndim)
+
+        # Map to reference cell
+        reference_cell = {x1: 0, y1: 0, z1: 0, x2: h, y2: h, z2: h}
```

Comment:
Yeah, the `h` assumption is hard-coded in many places atm, so picking it apart only here will have little effect. I'd hold off until we are building a proper `Grid` abstraction and can test the non-square/non-cube cases for stencils and point interpolation - hopefully this will be very soon.